src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs

```diff
@@ -965,20 +976,27 @@ private HRESULT SaveStringToHandle(IntPtr handle, string str, bool unicode)
                     return HRESULT.E_OUTOFMEMORY;
                 }
 
-                char[] chars = str.ToCharArray(0, str.Length);
-                UnsafeNativeMethods.CopyMemoryW(ptr, chars, chars.Length * 2);
+                int bytesToCopy = str.Length * 2;
+                fixed (char* pStr = str)
+                {
+                    Buffer.MemoryCopy(pStr, (char*)ptr, bytesToCopy, bytesToCopy);
+                }
+
+                // Add null terminator.
+                *(char*)IntPtr.Add(ptr, bytesToCopy) = '\0';
```

Comment:
nit: I think this can be simplified? Other instances as well, though be careful of the unit of the pointer (char or byte) if you do this refactoring!
```suggestion
                ptr[str.Length] = '\0';
```